anvil/src/state.rs

```diff
@@ -388,15 +390,83 @@ impl<BackendData: Backend + 'static> AnvilState<BackendData> {
         }
     }
 
-    pub fn send_frames(&self, output: &Output) {
-        self.space.elements().for_each(|window| {
-            if self.space.outputs_for_element(window).contains(output) {
-                window.send_frame(self.start_time.elapsed().as_millis() as u32)
+    pub fn send_frames(&self, output: &Output, report: &OutputRenderReport) {
+        let now = Instant::now();
+        let throttle = Some(Duration::from_secs(1));
+        let time = self.start_time.elapsed().as_millis() as u32;
+
+        for window in self.space.elements() {
+            let outputs = self.space.outputs_for_element(window);
+
+            // If the element outputs do not contain
+            // the requested output we can exit early
+            // as all further checks for the primary output
+            // will fail as well
+            if !outputs.contains(output) {
+                continue;
             }
-        });
+
+            window.send_frame(time, |surface, states| {
+                // We only want to send frame callbacks on the preferred output
+                if !window_surface_preferred_output(surface, &*outputs)
```

Comment:
Shouldn't the send_frame only occur after all outputs the surface is shown on have had their part of the surface presented?

Reply:
Why? That would limit the frame callbacks to the lowest refresh rate of all outputs it is shown.
The opposite is what you want most of the time, use the output with the highest refresh rate to drive the frame callbacks.

Reply:
Looking at this from a power efficiency standpoint, there's no point in starting the client's next draw until all parts of the last one have been displayed.

Suppose you have 2 outputs side by side where the window is spanned across both, one is 120Hz the other is 30Hz. Why would you kick off the next draw faster than 30Hz?

Reply:
Because 30 fps will look pretty bad on a 120Hz monitor? Or are least not so nice how it could look like, I mean what is the point in running a monitor with that refresh rate when you limit clients then? The output with the lower refresh rate would sample the buffers at it's refresh rate.
But it may be a valid point for your use case and this is why I struggle what should be put in smithay and what in anvil.

Reply:
Summing up IRC conversation. This is making a choice between power (and overall time where the windows remain in sync), and performance. For the sake of making progress, let's put a pin in it, since I think coverage is a good general heuristic. Perhaps in the future an argument can be added for providing a strategy:
coverage
slowest
fastest